src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,93 @@
+#include "quality_estimator.h"
+
+#include <iostream>
+
+#include "byte_array_util.h"
+
+namespace marian {
+namespace bergamot {
+
+QualityEstimator::QualityEstimator(AlignedMemory qualityEstimatorMemory) {
+  // QE coefficients memory passed as bytes.
+  QualityEstimator::load(std::move(qualityEstimatorMemory));
+}
+
+void QualityEstimator::load(AlignedMemory file_parameters) {
+  char delimiter = '\n';
+  int line_position = 0;
+  std::string line(file_parameters.begin());
+  std::vector<std::string> file_input;
+  file_input.reserve(4);
+  std::istringstream istr(line);
+  while (std::getline(istr, line, delimiter)) {
+    file_input.emplace_back(line);
+  }
+  ABORT_IF(file_input.size() != 4, "Model file should contains 4 lines, one per model parameter");
+
+  QualityEstimator::initVector(this->stds_, file_input[0]);
+  QualityEstimator::initVector(this->means_, file_input[1]);
+  QualityEstimator::initVector(this->coefficients_, file_input[2]);
+  QualityEstimator::initVector(this->intercept_, file_input[3]);
+}
+
+void QualityEstimator::initVector(std::vector<float>& emptyVector, std::string line) {
+  char delimiter = ' ';
+  size_t start = 0;
+  size_t end = line.find_first_of(delimiter);
+
+  while (end <= std::string::npos) {
+    emptyVector.push_back(std::stof(line.substr(start, end - start)));
```

Comment:
This parser is quite inefficient.  You're going to need a binary file format anyway, which will save us from arguing about this code.

Reply:
Hey @kpu, I refactored the initVector method, and I basically get rid of all previous code (2363e0a).

I'm already using AlignMemory instead of a text file, and if I understood the code correctly, it is a binary format already. If this is not what is the expected behavior that you want, could you please clarify?

Moreover, I also wrote an attempt to use the `features` for the quality estimator model. I'm not so sure if this would best implementation, so I'm open to suggestions here. The idea behind my implementation is to use an AlignedVector to represent the feature matrix because I would perform the matrix multiply efficiently with the model coefficients.

Thanks!

Reply:
You have a text format.  It has lines.  It has floats as strings.  

Let us count the number of times you're copying the file:
1. `std::string line(file_parameters.begin());`
2. `std::istringstream istr(line);`
3. `while (std::getline(istr, line, delimiter))` and `file_input.emplace_back(line);` count as one
4. `std::istringstream istr(line);` in `initVector`
5. `while (std::getline(istr, line, delimiter))` in `initVector`

You should have zero copying.  

Your parsing should look roughly like this, though I haven't bothered to separate it out into functions or class structure.  
```
struct FileHeader {
  uint64_t magic;
  /* Some of these may be redundant, please collapse */
  uint64_t std_size;
  uint64_t means_size;
  uint64_t coefficient_size;
  uint64_t intercept_size;
};
constexpr std::size_t kHeaderSize = 64 /* Keep alignment by being a multiple of 64 bytes */;
ABORT_IF(file_parameters.size() < kHeaderSize, "Quality estimation file too small");
const FileHeader &header = *reinterpret_cast<const FileHeader*>(file_parameters.begin());
constexpr uint64_t kMagic = 0x78cc336f1d54b180; /* Some random number to check file header */
ABORT_IF(kMagic != header.magic, "Incorrect magic bytes for quality estimation file");
ABORT_IF(file_parameters.size() != kHeaderSize + payloadSize(header), "Quality estimation file wrong size for claimed content");
const float *base = reinterpret_cast<const float*>(file_parameters.begin() + kHeaderSize);
stds_ = base;
means_ = (base += header.std_size);
coefficients_ = (base += header.means_size);
intercept_ = (base += header.coefficient_size);
```

Reply:
@kpu Here is the fix:  b52ee50

Hope it's fine now. Thanks.